interactive_engine/executor/ir/core/src/glogue/extend_step.rs

```diff
@@ -100,94 +172,62 @@ impl ExactExtendStep {
+    pub fn generate_expand_operators_vec(
+        &self, origin_pattern: &Pattern, is_intersect: bool,
+    ) -> IrPatternResult<Vec<Vec<pb::logical_plan::Operator>>> {
+        let mut expand_oprs_vec = vec![];
+        for extend_edge in self.iter() {
+            let edge_id = extend_edge.get_edge_id();
+            let edge_data = origin_pattern.get_edge_data(edge_id)?.clone();
+            let mut is_pure_path = false;
+            let mut expand_oprs = match edge_data {
+                PbEdgeOrPath::Edge(edge_opr) => {
+                    vec![extend_edge.generate_edge_expand(edge_opr)?]
+                }
+                PbEdgeOrPath::Path(path_opr) => {
+                    // For path expansion that doesn't have further intersection,
+                    // it is a pure path expansion
+                    is_pure_path = !is_intersect;
+                    extend_edge.generate_path_expands(path_opr, is_intersect)?
+                }
+            };
+            // every exapand should followed by an getV operator to close
+            let get_v =
+                self.generate_get_v_operator(origin_pattern, extend_edge.get_direction(), is_pure_path)?;
+            expand_oprs.push(get_v.clone());
+            expand_oprs_vec.push(expand_oprs);
         }
-
-        Ok(expand_operators)
+        Ok(expand_oprs_vec)
     }
 
     /// Generate the intersect operator for ExactExtendStep's target vertex
     /// It needs its parent EdgeExpand Operator's node ids
     pub fn generate_intersect_operator(
-        &self, parents: Vec<i32>,
+        &self, parents: Vec<KeyId>,
     ) -> IrPatternResult<pb::logical_plan::Operator> {
-        Ok((pb::Intersect { parents, key: Some((self.target_vertex_id as i32).into()) }).into())
+        Ok((pb::Intersect { parents, key: Some((self.target_vertex_id as KeyId).into()) }).into())
     }
 
-    /// Generate the filter operator for ExactExtendStep's target vertex if it has filters
-    pub fn generate_vertex_filter_operator(
-        &self, origin_pattern: &Pattern,
-    ) -> IrPatternResult<Option<pb::logical_plan::Operator>> {
-        // pick target vertex's property and predicate info from origin pattern
+    /// Generate a getV operator to close the outE
+    pub fn generate_get_v_operator(
+        &self, origin_pattern: &Pattern, expand_direction: PatternDirection, is_pure_path: bool,
+    ) -> IrPatternResult<pb::logical_plan::Operator> {
         let target_v_id = self.target_vertex_id;
-        if let Some(vertex_data) = origin_pattern.get_vertex_data(target_v_id) {
-            if vertex_data.predicate.is_some() {
-                return Ok(Some(vertex_data.clone().into()));
+        let vertex_params = origin_pattern
+            .get_vertex_parameters(target_v_id)?
+            .cloned();
+        // when meet pure path expand, the VOpt should always be EndV
+        let get_v_opt = if is_pure_path {
+            pb::get_v::VOpt::End as i32
```

Comment:
Seems a misunderstanding here. EndV for path do not carry any filters. Instead, the GetV insides the ExpandBase of path would.